sts-mock/src/token/serviceTokenGenerator/serviceTokenGenerator.ts

```diff
@@ -0,0 +1,126 @@
+import { errorResult, Result, successResult } from "../../utils/result";
+import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
+import { importJWK, JWK, JWTPayload, SignJWT, KeyLike } from "jose";
+
+export class ServiceTokenGenerator implements IServiceToken {
+  private readonly stsMockBaseUrl: string;
+  private readonly keyStorageBucket: string;
+  private readonly subjectId: string;
+  private readonly scope: string;
+  private readonly tokenTimeToLive: number;
+  private s3Client = new S3Client([
+    {
+      region: "eu-west-2",
+      maxAttempts: 2,
+    },
+  ]);
+
+  constructor(
+    stsMockBaseUrl: string,
+    keyStorageBucket: string,
+    subjectId: string,
+    scope: string,
+    tokenTimeToLive: number,
+  ) {
+    this.stsMockBaseUrl = stsMockBaseUrl;
+    this.keyStorageBucket = keyStorageBucket;
+    this.subjectId = subjectId;
+    this.scope = scope;
+    this.tokenTimeToLive = tokenTimeToLive;
+  }
+
+  async generateServiceToken(): Promise<Result<string>> {
+    const getPrivateKeyFromS3Result = await this.getPrivateKeyFromS3();
+    if (getPrivateKeyFromS3Result.isError) {
+      return getPrivateKeyFromS3Result;
+    }
+
+    const formatKeyAsKeyLikeResult = await this.formatKeyAsKeyLike(
+      getPrivateKeyFromS3Result.value,
+    );
+    if (formatKeyAsKeyLikeResult.isError) {
+      return formatKeyAsKeyLikeResult;
+    }
+
+    const tokenPayload = this.createServiceTokenPayload();
+
+    const signingResult = await this.signServiceToken(
+      tokenPayload,
+      formatKeyAsKeyLikeResult.value,
+    );
+    if (signingResult.isError) {
+      return signingResult;
+    }
+
+    return successResult(signingResult.value);
+  }
+
+  private async getPrivateKeyFromS3(): Promise<Result<string>> {
+    const retrieveParams = {
+      Bucket: this.keyStorageBucket,
+      Key: "private-key.json",
+    };
+    try {
+      const response = await this.s3Client.send(
+        new GetObjectCommand(retrieveParams),
+      );
+      return successResult(await response.Body!.transformToString());
+    } catch {
+      return errorResult({
+        errorMessage: "Unable to fetch file from S3",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+  }
+
+  private async formatKeyAsKeyLike(
```

Comment:
Is it possible to rename this to something specific? I don't have a good suggestion as I've not looked at what the importJWK func does yet

Reply:
Or is key like an actual term that's commonly used? If so then I wouldn't see this as a problem

Reply:
Ah I can see it's a type from Jose. Maybe this is fine then!